Integrate VeSync humidifiers

jspayd opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
- Integrates VeSync humidifiers
- Upgrades pyvesync to version 2.0.1. There is no changelog message, but upgrading allows us to look up what features humidifiers support. E.g., we can tell if humidifiers support a night light just by checking for the `night_light` property.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://github.com/home-assistant/core/issues/44865 (closed, but only due to inactivity)
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]: https://github.com/home-assistant/home-assistant.io/pull/20901

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description. - This is the latest changelog message. Most of the updates don't have changelog messages.
   > Adding support for the Core300S air purifier and fix Azure Pipelines (https://github.com/webdjoe/pyvesync/releases/tag/1.4.2)


- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.
   - https://github.com/home-assistant/core/pull/62904#pullrequestreview-840623656
   - https://github.com/home-assistant/core/pull/62900#pullrequestreview-840624010

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jspayd committed:
Integrate VeSync Classic300S humidifiers

homeassistant commented:
Hi @jspayd, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#62907).

Once you do that we will be able to review and accept this pull request.

Thanks!

decompil3d commented on homeassistant/components/vesync/humidifier.py:
```diff
@@ -0,0 +1,155 @@
+
+        if "humidity" in self.device.details:
+            # The "humidity" key is reserved by HumidifierEntity for target humidity
+            attr["current_humidity"] = self.device.details["humidity"]
```
Consider exposing this as a separate sensor instead of an attribute. My PR #61837 adds device registrations for devices in this component, which would allow you to tie together the humidifier and the humidity sensor into a single device.

jspayd replied:
I'm pretty new to writing Home Assistant integrations. Are entities always considered strictly better than state attributes that can be turned into entities using templates? Are there guidelines for deciding what should be an entity and what should be a state attribute? Instead of making less-commonly used properties state attributes, is it not better to make them entities tagged with `entity_registry_enabled_default=False`?

The reason I'm asking is that I wonder if I should turn the other state attributes into separate entities as well.

decompil3d replied:
I think about it in terms of use case -- how likely is it that someone will want to use the information? If it's just simple diagnostic data that's infrequently referenced, attributes make sense. But I feel like a humidity sensor is pretty useful top-level information to expose.

But I am no expert either -- only a handful of PRs merged so far 😉 I'd love to hear from the maintainers on this.

jspayd replied:
Funny enough, diagnostic tools are precisely one of the examples provided of what should be an entity that's disabled by default (see https://developers.home-assistant.io/docs/core/entity/#advanced-properties). The guidance is certainly nebulous, so I'd love to hear from the maintainers as well.

decompil3d commented on homeassistant/components/vesync/number.py:
```diff
@@ -0,0 +1,114 @@
+MIN_MIST_LEVEL = 1
+MIST_LEVEL_STEP = 1
+
+MAX_NIGHT_LIGHT_BRIGHTNESS = 100
```
Can we expose the night light as a Light entity tied to the humidifier device rather than as a Number? Allows use of existing brightness support.

jspayd replied:
Sure, that makes sense.

decompil3d commented on homeassistant/components/vesync/number.py:
```diff
@@ -0,0 +1,114 @@
+class VeSyncHumidifierNightLightHA(VeSyncDevice, NumberEntity):
+    """Representation of the night light on a VeSync humidifier.
+
+    This is a number entity instead of a light entity because it does not
+    support an on/off toggle.
```
I assume setting brightness to 0 turns it off and >0 turns it on? Could you just treat an `on` command as setting brightness to 100 and `off` as setting to 0?

jspayd replied:
Sure

decompil3d commented on homeassistant/components/vesync/switch.py:
```diff
@@ -106,3 +110,57 @@ def __init__(self, switch):
+        """Return the ID of this display."""
+        return f"{super().unique_id}.display"
+
+    @property
```
Mark entity as `config` (see docs under `entity_category` [here](https://developers.home-assistant.io/docs/core/entity#generic-properties))

```suggestion
    @property
    def entity_category(self):
        """Return the config entity category."""
        return EntityCategory.CONFIG
    
    @property
```

decompil3d commented on homeassistant/components/vesync/switch.py:
```diff
@@ -106,3 +110,57 @@ def __init__(self, switch):
+        """Return the ID of this device."""
+        return f"{super().unique_id}.automatic_stop"
+
+    @property
```
Mark entity as `config` (see docs under `entity_category` [here](https://developers.home-assistant.io/docs/core/entity#generic-properties))

```suggestion
    @property
    def entity_category(self):
        """Return the config entity category."""
        return EntityCategory.CONFIG
    
    @property
```

jspayd replied:
Thanks for the suggestion! I'll do that.

decompil3d commented on homeassistant/components/vesync/switch.py:
```diff
@@ -5,7 +5,7 @@
 from homeassistant.core import callback
 from homeassistant.helpers.dispatcher import async_dispatcher_connect
```
```suggestion
from homeassistant.helpers.dispatcher import async_dispatcher_connect
from homeassistant.helpers.entity import EntityCategory
```

decompil3d commented on homeassistant/components/vesync/number.py:
```diff
@@ -0,0 +1,114 @@
+
+from homeassistant.components.number import NumberEntity
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
```
```suggestion
from homeassistant.helpers.dispatcher import async_dispatcher_connect
from homeassistant.helpers.entity import EntityCategory
```

decompil3d commented on homeassistant/components/vesync/number.py:
```diff
@@ -0,0 +1,114 @@
+        """Return the ID of this device."""
+        return f"{super().unique_id}.night_light"
+
+    @property
```
Mark entity as `config` (see docs under `entity_category` [here](https://developers.home-assistant.io/docs/core/entity#generic-properties))

```suggestion
    @property
    def entity_category(self):
        """Return the config entity category."""
        return EntityCategory.CONFIG
    
    @property
```

decompil3d commented on homeassistant/components/vesync/number.py:
```diff
@@ -0,0 +1,114 @@
+        """Return the ID of this device."""
+        return f"{super().unique_id}.mist_level"
+
+    @property
```
Mark entity as `config` (see docs under `entity_category` [here](https://developers.home-assistant.io/docs/core/entity#generic-properties))

```suggestion
    @property
    def entity_category(self):
        """Return the config entity category."""
        return EntityCategory.CONFIG
    
    @property
```

decompil3d reviewed (changes requested):
Thanks for your contribution! I have an open PR for this component, so I tried to note places where we can align to avoid merge conflicts later on.

jspayd commented on homeassistant/components/vesync/humidifier.py:
```diff
@@ -0,0 +1,155 @@
+    for dev in devices:
+        if DEV_TYPE_TO_HA.get(dev.device_type) == "humidifier":
+            dev_list.append(VeSyncHumidifierHA(dev))
+        else:
+            continue
```
Agreed. I think this was a copy-paste from another file, but I can fix it here.

jspayd commented on homeassistant/components/vesync/common.py:
```diff
@@ -40,7 +51,7 @@ async def async_process_devices(hass, manager):
 
 
-class VeSyncDevice(ToggleEntity):
+class VeSyncDevice:
```
Do you expect your PR to be merged soon? If so, I could just wait for that before making any changes.

decompil3d replied:
Not certain, but AFAIK it's on track for the February release.

jspayd commented:
Considering there is another open PR for this integration (https://github.com/home-assistant/core/pull/61837), I'm going to wait a while before making the requested changes. I'm hoping that the other PR is merged soon so that I can rebase and *then* make changes, since that should be easier than trying to make the merge clean by eye and then still having to do some work when I rebase.

jspayd commented:
Other PR was merged, so I rebased and made the recommended changes. Still need to update my docs PR.

jspayd commented:
Updated docs

decompil3d commented on homeassistant/components/vesync/sensor.py:
```diff
@@ -150,3 +151,51 @@ def update(self):
+
+    @property
+    def native_value(self):
+        """Return the current power usage in W."""
```
```suggestion
        """Return the current humidity in percent."""
```

decompil3d reviewed (approved):
Tiny nit on a doc comment but otherwise LGTM. Thanks for waiting on my other PR and integrating really nicely with those changes.

jspayd committed:
Fix docstring for VeSyncHumiditySensor.native_value

Co-authored-by: Jonathan Keslin <decompil3d@users.noreply.github.com>

jspayd commented:
> Tiny nit on a doc comment but otherwise LGTM. Thanks for waiting on my other PR and integrating really nicely with those changes.

Fixed. Thanks for reviewing!

heebnerd commented on homeassistant/components/vesync/const.py:
```diff
@@ -9,4 +9,25 @@
+    "Core200S": "fan",
+    "Core300S": "fan",
+    "Core400S": "fan",
+    "Classic300S": "humidifier",
```
I think you could also add "Dual200S": "humidifier", to this list

jspayd replied:
While it looks like the master branch of pyvesync has Dual200S support, the PR for that was just merged two weeks ago (https://github.com/webdjoe/pyvesync/pull/104), so it's not in the latest release, which is from November. This should be possible after the next pyvesync release, though.

borpin replied:
I have found my Humidifier, ostensibly a Dual200S is reporting as "LUH-D301S-WEU" so it would be good to add that model as well.

From pyvsync:
```
{
	"Device Name": "Study Hum",
	"Model": "LUH-D301S-WEU",
	"Subdevice No": "None",
	"Status": "on",
	"Online": "online",
	"Type": "wifi-air",
	"CID": "vsaqxxxxxxxxxxxxxxxxxxxxxxx",
	"Mode": "auto",
	"Humidity": "40",
	"Mist Virtual Level": "1",
	"Mist Level": "1",
	"Water Lacks": false,
	"Humidity High": false,
	"Water Tank Lifted": false,
	"Display": true,
	"Automatic Stop Reach Target": true,
	"Auto Target Humidity": "42",
	"Automatic Stop": true
}
````

jspayd committed:
Merge branch 'dev' of https://github.com/home-assistant/core into vesync-humidifier

jspayd committed:
Merge branch 'vesync-humidifier' of github.com:jspayd/core into vesync-humidifier

jspayd committed:
Add type hints to VeSync humidifier.py & number.py

johndbritton commented:
What does this PR need to be merged? Anything I can do to help?

jspayd commented:
> What does this PR need to be merged? Anything I can do to help?

I think it just needs review by a maintainer. This is my first Home Assistant PR, so I'm not completely sure how long that normally takes or what the process is like.

decompil3d commented:
The maintainers often seem to review PRs in bursts, often right before a new release. So I'd expect reviews next week ahead of the 2/2 release.

alexanv1 commented:
@jspayd, the Humidifier also includes a couple of sensors that are generally useful ("out of water" and "removed from base"). I've added support for those in a forked branch (https://github.com/alexanv1/core/tree/vesync_humidifier_support). Unfortunately I completely messed up when merging so only have a single commit with both your changes and new binary_sensor additions.

Do you want to include the binary sensors in this PR?
If not, I will do a follow-up PR to add those once this is merged.

ekinnee commented:
New release of pyvesync.

jspayd commented:
@alexanv1 I think it would make sense to include your changes in this PR. That way we don't risk breaking backwards compatibility in the event that there's a new Home Assistant release between the time that my PR gets merged and when your PR gets merged. And you should be able rebase off of my changes to clean up the commit history. After I upgrade this to the new pyvesync release and add the Classic200S, could you make a PR to my `vesync-humidifier` branch, and also make another PR to my docs branch (see https://github.com/home-assistant/home-assistant.io)?

jspayd commented:
Blocked on adding the Classic 200S humidifiers until the display toggle is fixed in pyvesync: https://github.com/webdjoe/pyvesync/issues/114

jspayd commented:
> Do we really need to hold up 300 support due to 200 issues?

We don’t. It’s 200S support that’s blocked, not this PR. The 200S can come in a later PR if necessary. Another option would be to add support for the 200S without the display toggle.

@alexanv1 Since 200S support is blocked, you can disregard my previous comment about waiting for 200S support before PR-ing your changes, though it will also be fine if it’s in a later PR if you don’t get to it before this one is merged.

alexanv1 commented:
> @alexanv1 Since 200S support is blocked, you can disregard my previous comment about waiting for 200S support before PR-ing your changes, though it will also be fine if it’s in a later PR if you don’t get to it before this one is merged.

@jspayd, I will try to submit PRs later today or tomorrow.
However, there is another issue that I think might prevent this from being merged as-is... I have 2 humidifiers and ran into a problem with VeSync APIs failing on a regular basis. The most likely reason is rate-limiting and I believe it's caused by every entity triggering an update independently via VeSync APIs. I worked around this in my private branch by making the update no-op on everything but the "main" entity (i.e. the humidifier entity in this case). Other entities then make use of the cached data.

However, the way I did it is a hack and it probably requires an update coordinator or a similar pattern but I am not familiar with how that works at all.

jspayd commented:
@alexanv1 Thanks for the info. I also have 2 humidifiers and did not run into the same issue, though I didn't run Home Assistant with both of them added for very long. I'll take some time to try to reproduce it, or if you're able to find an error message that confirms it's rate limiting, that would be helpful.

If necessary, I can look into an update coordinator, though that will take a bit of time since I have 0 experience developing for Home Assistant before this PR and I am also unfamiliar with the update coordinator pattern.

alexanv1 commented:
> @alexanv1 Thanks for the info. I also have 2 humidifiers and did not run into the same issue, though I didn't run Home Assistant with both of them added for very long. I'll take some time to try to reproduce it, or if you're able to find an error message that confirms it's rate limiting, that would be helpful.
> 

The errors look like the below:
```
2022-02-14 16:23:17 ERROR (MainThread) [homeassistant.helpers.entity] Update for humidifier.nathan_s_bedroom_humidifier fails
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 521, in async_update_ha_state
    await self.async_device_update()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 725, in async_device_update
    raise exc
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/vesync/common.py", line 125, in update
    self.device.update()
  File "/usr/local/lib/python3.9/site-packages/pyvesync/vesyncfan.py", line 1183, in update
    self.get_details()
  File "/usr/local/lib/python3.9/site-packages/pyvesync/vesyncfan.py", line 1164, in get_details
    outer_result = r.get('result', {})
AttributeError: 'NoneType' object has no attribute 'get'
```

There is nothing explicit in the error about rate limiting - perhaps it's being masked by the underlying pyvesync library. I still see the errors sometimes but I was getting them constantly (100s of times a day in the logs) prior to making the changes to no-op the updates on "secondary" entities. Of note, because the errors are on a specific entity update, they are mostly unnoticeable except in the logs - since subsequent updates usually succeed.

NickM-27 commented:
I have two humidifiers (200s and 300s) and have not seen any issues like above. 

Also, really glad to see this making its way into the main HASS component as right now I am running a fork as a custom_component.

kevchu3 commented:
Hi @jspayd I'm looking at your PR and it looks great.  I've just written and opened a PR for an incompatible version with yours, and wish I came across yours sooner:
- https://github.com/home-assistant/core/pull/66683

My PR originated from trying to solve this issue:
- https://github.com/home-assistant/core/issues/61417#issuecomment-998816504

And the design question that we should ask is whether we want to treat fans and humidifiers the same or differently in HA core.  The maintainers for [pyvesync](https://github.com/webdjoe/pyvesync) treat both fans and humidifiers alike (see: [vesyncfan.py](https://github.com/webdjoe/pyvesync/blob/master/src/pyvesync/vesyncfan.py)).  However, here in HA core, there are unimplemented functions and variables in humidifiers if we strictly treat them as fans today (which is why I see you've implemented a separate humidifier type).  And thus, this is a design question for the maintainers (@markperdue, @webdjoe, @TheGardenMonkey)

kevchu3 reviewed (commented):
Hi @jspayd I'm looking at your PR and it looks great.  I've just written and opened a PR for an incompatible version with yours, and wish I came across yours sooner:
- https://github.com/home-assistant/core/pull/66683

My PR originated from trying to solve this issue:
- https://github.com/home-assistant/core/issues/61417#issuecomment-998816504

And the design question that we should ask is whether we want to treat fans and humidifiers the same or differently in HA core.  The maintainers for [pyvesync](https://github.com/webdjoe/pyvesync) treat both fans and humidifiers alike (see: [vesyncfan.py](https://github.com/webdjoe/pyvesync/blob/master/src/pyvesync/vesyncfan.py)).  However, here in HA core, there are unimplemented functions and variables in humidifiers if we strictly treat them as fans today (which is why I see you've implemented a separate humidifier type).  And thus, this is a design question for the maintainers (@markperdue, @webdjoe, @TheGardenMonkey)

jspayd commented:
After running with two humidifiers for a few days, I do see the error that @alexanv1 brought up in the logs. Unfortunately the error from pyvesync is pretty opaque, and I don't want to just assume it's rate limiting. I'll see if I can get visibility into the full API response somehow.

jspayd commented:
Ah, looks like a read timeout:
```
2022-02-17 22:57:08 WARNING (SyncWorker_5) [pyvesync.helpers] HTTPSConnectionPool(host='smartapi.vesync.com', port=443): Read timed out. (read timeout=5)
```

If it were rate limiting, I'd expect to receive a response indicating that that's the case, but the server isn't sending a response at all. I suppose it is possible that VeSync rate limits by not responding to some requests, but I'm not seeing enough to convince me that the issue is on the client end.

jspayd commented:
Considering we don't have any visibility into why the server isn't responding to some requests, and because it's probably good practice anyway, I went ahead and implemented an update coordinator for VeSync devices. After a day of running, there are still server errors. There are fewer of them, but that can probably be explained just by the fact that there are fewer requests to the server. Unfortunately I don't have any VeSync devices with existing Home Assistant integrations to see if the error appears for them too. The errors don't appear to result in a degraded user experience, so I'm not sure how concerning they are.

I'm not adding the update coordinator to the PR yet since I'm having trouble getting it to update right after a setting is changed instead of waiting until the scheduled poll.

decompil3d commented:
FWIW, I just added a vesync air purifier to my set of vesync outlets and now see this error a bunch. And outlets have gotten really flaky on updates -- I often have to trigger on/off operations multiple times to make things work. So getting a `DataUpdateCoordinator` is probably a good next step to address this. Could even go into a separate PR since this api error occurs without your changes.

jspayd commented:
I don’t think adding an update coordinator is going to fix this, since I implemented one and, while it was buggy (updates didn’t happen right after, e.g. flipping a switch) it did succeed in collapsing API requests, and the error still occurred. But I agree that since this error is not unique to humidifiers, and since in my experience it doesn’t result in a degraded user experience, it can come in a separate PR. Actually it probably requires a change to pyvesync that increases timeouts and/or adds retries, since it’s caused by request timeouts. Since this is happening with fans but not outlets (?), perhaps some devices just have greater API latency.

jspayd commented:
To anyone reviewing this: This PR should be in a state where it can be merged. A recent pyvesync change fixes the bug with the Classic200S, so once that’s released I can add support for the 200S as well, but I can do that in a separate (very small) PR if that doesn’t happen before this one is merged.

webdjoe commented:
The only way I can see to address this is to catch the error gracefully. Without knowing the exact rate limit or reason for the timeouts, I don't think this library would be the best place to include update delays. The end-user can implement delays on their end as it suits their needs. 

Vesync has been hardening its API, using multiple, very effective certificate pinning processes on the android app. They designed it only to be updated while using the app, while HA updates continuously increasing their server load. They already changed the login and device update api, which isn't updated yet in this library. It's only a matter of time before they depreciate it entirely and force users to update the app.

borpin commented on homeassistant/components/vesync/humidifier.py:
```diff
@@ -0,0 +1,161 @@
+HUMIDIFIER_MODE_SLEEP = "sleep"
+
+AVAILABLE_MODES = {
+    "Classic300S": [
```
Requires other models to be added here such as `LUH-D301S-WEU` and `Dual200S`

jspayd replied:
I won't be able to test those since I don't have them. Do you still want me to add them?

borpin commented:
@jspayd Looking at this, the base library uses dicts for the different models. Could this integration simply pull in those dicts rather than recreate them locally for modes and features etc? 

https://github.com/webdjoe/pyvesync/blob/c41eb8a77e68cc357f558e4fa413e42f9ba68658/src/pyvesync/vesyncfan.py#L10-L10

This would then mean (I think) that when a new model is added to the base library, simply updating the manifest, should bring that model in. If additional features are then made available, additional work might then be required.

As it is, each model in each group needs to be incorporated along with all their features.

As it stands, things like Mist levels also still need to be implemented as they are fixed for the one model of humidifier currently identified.

This could cover the other open PRs for this integration as well.

jspayd committed:
Read humidifier and feature lists from pyvesync

Includes upgrade to pyvesync 2.0.0

jspayd commented:
Ok this latest update
1. upgrades to pyvesync 2.0.0;
2. reads the list of humidifiers from pyvesync; and
3. reads max/min mist_level and supported modes from pyvesync.

This means that every humidifier supported by pyvesync should be supported by this version of of my PR, though I can only test the Classic300S and Classc200S.

The only issue is that pyvesync 2.0.0 has a bug that breaks the night light for the Classic300S, and also marks the Classic200S as having a nightlight when it does not (see https://github.com/webdjoe/pyvesync/issues/125). These are simple to fix, so I expect they'll be fixed soon.

I also need to update my docs PR with info on the new humidifier models, which I should have time to do some time this week.

borpin commented:
> This means that every humidifier supported by pyvesync should be supported by this version of of my PR

Great 😄. 

> I can only test the Classic300S and Classc200S.

I had to go through hoops to be able to use this PR. Is there a 'simple' way to do it (running HAOS and cannot get the SSH keys thing to work)? Otherwise I'll have to do my cutting and pasting again to pull it in as a custom-component 😦 

> has a bug that breaks the night light for the Classic300S

I think it might be a typo. https://github.com/webdjoe/pyvesync/issues/125

While here, could you add some more information as sensors rather than just attributes?

- water lacks (a change in state of this will trigger an alert to fill it)
- tank lifted (it will also tell you if there is a blockage during use)
- humidity high (a change in state might trigger a fan to be started)

Max and min humidity attributes are not relevant to my humidifier - just wonder what their purpose is?

borpin commented:
> I can only test the Classic300S and Classc200S.

I am getting the unknown device error again.

```
WARNING (MainThread) [custom_components.vesync2.fan] Study Hum - Unknown device type - LUH-D301S-WEU
```

This is one of the 'sub models' in the dict -

https://github.com/webdjoe/pyvesync/blob/c41eb8a77e68cc357f558e4fa413e42f9ba68658/src/pyvesync/vesyncfan.py#L25

I had to add it as a specific model before.

decompil3d commented:
Folks, this PR is getting a lot of scope creep. Can we work to get it into a reasonable mergeable state and then open follow-up PRs for additional features/entities/devices?

borpin commented:
> Can we work to get it into a reasonable mergeable state and then open follow-up PRs for additional features/entities/devices?

As it stands it works for me bar the issue with the step above. It could perhaps do with a title change, but it will now work for any device in the Python Library and that is a major win.  The binary sensors will be easily added if not included. I just wanted to use them 😆.

borpin commented:
> can tell you that the LV600S also works

Great 😄 

> but it only lacks the heat component (yup, it has an option to send out warm mist)

The integration needs a new entity to be created where this feature exists. I suggest we probably need to leave that for now. It is listed in the Library as a feature, so picking that up should not be too much of an issue.

jspayd commented:
Gathering the issues I see listed for my own benefit:
1. mist level step (I think I saw in docs the default was 1 and let it use the default, but it looks like it ended up not being 1? I'll have to take a closer look)
2. device type lookup is wrong
3. night light issues (fixed in latest pyvesync release)

These should all be easy fixes. Working on it now.

jspayd committed:
Upgrade to pyvesync 2.0.1 and fix vesync humidifier bugs

jspayd commented:
Fixed the issues listed above and updated my docs PR. I'm very much in favor of not adding any more functionality in this PR, and allowing people to add functionality in later PRs.

jspayd commented:
> Max and min humidity attributes are not relevant to my humidifier - just wonder what their purpose is?

The Classic300S and Classic200S support target humidities of 30-80%. It looks like pyvesync doesn't keep a record of what the max/min humidities are for each model. Perhaps that's something that can be added in a later PR.

borpin commented:
> 1. mist level step (I think I saw in docs the default was 1 and let it use the default, but it looks like it ended up not being 1? I'll have to take a closer look)

https://developers.home-assistant.io/docs/core/entity/number If the difference is one, the step will be 0.1.

@jspayd,  @vlebourl has done a whole load of work on this via custom integration which started from this PR. You might want to take a look. https://github.com/vlebourl/custom_vesync

Also, the Python library has been bumped to fix a couple of bugs.

jakeasmith commented:
Heya! Just jumping in with another “this is gonna be awesome!” Any chance of this making it into the next release?

borpin commented:
> Heya! Just jumping in with another “this is gonna be awesome!” Any chance of this making it into the next release?

You can install an updated version via HACS (will just work instead of the core version). https://github.com/vlebourl/custom_vesync

Will work better even when this PR eventually gets released to core.

NickM-27 commented:
Bump

2opremio commented:
What's left for this to be merged? (apart from fixing the conflicts). I would really benefit from it (and @jspayd put a whole lot of effort into it).

borpin commented:
> I would really benefit from it

Don't hold your breath. I'd just use the custom integration https://github.com/vlebourl/custom_vesync

cdnninja commented:
I am going to try bump this back up.  @jspayd  Any chance you can resolve the merge conflicts as a start?   I can give this a test soon to confirm still working.

chuckdeal97 commented:
I'm looking for some help, I don't have a strong git background, and I'm not sure how to proceed.

I was following an article I found to add to someone else's pull request (https://tighten.com/insights/adding-commits-to-a-pull-request/) and did a checkout of jspayd's branch.  I then followed the steps described at https://developers.home-assistant.io/docs/development_catching_up/ for rebasing.  That branch was MANY months out-of-date and there were a number of conflicts and underlying code improvements.  I tried the final step of pushing the rebased changes to jspayd's branch but received permission denied.

> ➜  hass-core git:(jspayd-vesync-humidifier) git push jspayd HEAD:vesync-humidifier --force-with-lease
remote: Permission to jspayd/core.git denied to chuckdeal97.
fatal: unable to access 'https://github.com/jspayd/core.git/': The requested URL returned error: 403

I now have my own branch (https://github.com/chuckdeal97/hass-core/tree/jspayd-vesync-humidifier) that I am not sure how to use to push this PR ahead.

decompil3d commented:
You can't push to her branch but you can set the base of your PR to her branch and open it against her fork. Or you can just open a fresh PR against the upstream (here) and note the PR number here so the maintainers can close this one if appropriate.

chuckdeal97 commented:
For those following this PR, I'm attempting to merge the base content from the PR with PR #84025 .

The active PR for supporting humidifiers is #88941

chuckdeal97 commented:
I'd recommend closing this PR in favor of a newer set of PRs that have been opened to cover the same functionality (#90455)

epenet reviewed (changes requested):
There are merge conflicts that need resolving.
Please rebase or consider closing in favor of https://github.com/home-assistant/core/pull/90455
